sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -65,9 +64,43 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
 }
 
 /**
+ * Use StringType for expressions supporting only binary collation.
+ *
  * @since 1.3.0
  */
 @Stable
 case object StringType extends StringType(0) {
   def apply(collationId: Int): StringType = new StringType(collationId)
+  override private[sql] def acceptsType(other: DataType): Boolean =
+    other.isInstanceOf[StringType] && other.asInstanceOf[StringType].isDefaultCollation
```

Comment:
Is this needed? The default implementation should work well as it requires the collation id to be same as the current one, which is 0 (`case object StringType extends StringType(0)`).